### PR TITLE
[MDB IGNORE] Create Ship Console subtype

### PR DESCRIPTION
**Pull request by ZephyrTFA**
## About The Pull Request

Tired of struggling to get your consoles to have a reference to the simulated ship they're on?
Tired of them failing to link to their simulated ship and randomly lose connection to it?
Introducing the ship console subtype!

## Why It's Good For The Game

refactor good

## Changelog
:cl:
code: add new ship console subtype
refactor: helm console is now a subtype of a ship console
/:cl:

### CODE REVIEW

File: code/modules/overmap/ship_consoles/_ship_console.dm [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+	return length(SStgui.open_uis_by_src[our_ref])
+
+/obj/machinery/computer/ship/proc/update_state()
+	SHOULD_NOT_SLEEP(TRUE)
+	var/toggle = get_active_uis()
+	var/prev = activated
+	if(!!activated != !!toggle) // If we are not in the state we should be in
+		activated = !activated
+	if(activated != prev)
+		playsound(src, activated ? activate_sound : deactivate_sound, 50)
```
Comment (MarkSuckerberg) [inline code/modules/overmap/ship_consoles/_ship_console.dm]: Is this is better than the old way of doing it? ~~Genuinely asking, I am not really in a position to check~~

Comment (MarkSuckerberg) [reply code/modules/overmap/ship_consoles/_ship_console.dm]: Oh, also, does it account for UIs opened by ghosts

Comment (ZephyrTFA) [reply code/modules/overmap/ship_consoles/_ship_console.dm]: It doesn't really have the ability, without shitty snowflake code, to distingush between ghosts and mob/living unfortunately

Comment (Sun-Soaked) [reply code/modules/overmap/ship_consoles/_ship_console.dm]: "Captain, why is our vessel accelerating towards the sun?"

Comment (ZephyrTFA) [reply code/modules/overmap/ship_consoles/_ship_console.dm]: Ghosts can't interact with the console, only have it open. That won't be an issue

File: code/modules/overmap/ship_consoles/_ship_console.dm [hunk 2]
```diff
@@ -10,10 +10,13 @@
...
 
-/obj/machinery/computer/ship/proc/get_active_uis()
+/obj/machinery/computer/ship/proc/get_active_uis(living_only = TRUE)
 	PRIVATE_PROC(TRUE)
 	var/our_ref = "[REF(src)]"
-	return length(SStgui.open_uis_by_src[our_ref])
+	. = 0
+	for(var/datum/tgui/ui as anything in SStgui.open_uis_by_src[our_ref])
+		if(living_only && isliving(ui.user))
+			.++
```
Comment (ZephyrTFA) [inline code/modules/overmap/ship_consoles/_ship_console.dm]:
```suggestion
		if(living_only && !isliving(ui.user))
			continue
		.++
```

File: code/modules/overmap/ship_consoles/_ship_console.dm [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+	. = ..()
+	if(!attempt_connect())
+		message_admins("[src] failed to connect to a simulated ship!")
+
+/obj/machinery/computer/ship/proc/attempt_connect()
+	. = FALSE
+	if(simple_connect())
+		. = TRUE
+	if(recursive_connect())
+		. = TRUE
```
Comment (ZephyrTFA) [inline code/modules/overmap/ship_consoles/_ship_console.dm]:
```suggestion
	if(simple_connect())
		. = TRUE
	if(!. && recursive_connect())
		. = TRUE
```

**Comment by MarkSuckerberg** [comment]
Seemed to break on the one round it was testmerged, I only saw multiple admin messages about not being able to connect but I bet that meant the helm consoles didn't work at all

**Comment by ZephyrTFA** [comment]
@MarkSuckerberg This is now ready for a TM again, resolved the issue

**Comment by ZephyrTFA** [comment]
Mark 🦎 please

**Comment by MarkSuckerberg** [comment]
Ouaaa, I'll make sure to get it TM'd next uptime

File: code/modules/overmap/ship_consoles/_ship_console.dm [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+/// Recurse all simulated ships to identify the ship we reside in, possibly laggy so use sparingly
+/obj/machinery/computer/ship/proc/recursive_connect()
+	var/area/our_area = get_area(src)
+	for(var/obj/structure/overmap/ship/simulated/ship as anything in SSovermap.simulated_ships)
+		CHECK_TICK
+		for(var/area/area as anything in ship.shuttle.shuttle_areas)
+			if(our_area == area)
+				current_ship = ship
+				return TRUE
```
Comment (MarkSuckerberg) [inline code/modules/overmap/ship_consoles/_ship_console.dm]:
```suggestion
	var/area/ship/our_area = get_area(src)
	if(!istype(our_area))
		return FALSE
	current_ship = our_area.mobile_port.current_ship
	return TRUE
```
Honestly this is less expensive than proc/simple_connect like this

**Comment by ZephyrTFA** [comment]
@MarkSuckerberg is it worth maintaining this PR

**Comment by MarkSuckerberg** [comment]
I'm not sure it is, I had planned on just adding this sort of thing to a machinery-level var rather than just consoles, and I had a way to more performantly and reliably connect them to overmap objects.

**Pull request closed without merging**